File: src/weaponsprocurement/gui/StockReviewPendingTrades.java
package weaponsprocurement.gui;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class StockReviewPendingTrades {
    private final List<StockReviewPendingPurchase> trades = new ArrayList<StockReviewPendingPurchase>();

    List<StockReviewPendingPurchase> asList() {
        return Collections.unmodifiableList(trades);
    }

    boolean isEmpty() {
        return trades.isEmpty();
    }

    void clear() {
        trades.clear();
    }

    void replaceWith(List<StockReviewPendingPurchase> source) {
        trades.clear();
        if (source == null) {
            return;
        }
        for (int i = 0; i < source.size(); i++) {
            StockReviewPendingPurchase purchase = source.get(i);
            if (purchase != null && !purchase.isZero()) {
                trades.add(purchase.copy());
            }
        }
    }

    void add(String itemKey, String submarketId, int quantity) {
        if (itemKey == null || itemKey.isEmpty() || quantity == 0) {
            return;
        }
        StockReviewPendingPurchase existing = find(itemKey, submarketId);
        if (existing == null) {
            trades.add(new StockReviewPendingPurchase(itemKey, submarketId, quantity));
            return;
        }
        existing.addQuantity(quantity);
        if (existing.isZero()) {
            trades.remove(existing);
        }
    }

    void adjustItemNet(String itemKey, int delta) {
        if (itemKey == null || itemKey.isEmpty() || delta == 0) {
            return;
        }
        int remaining = delta;
        if (delta < 0) {
            remaining = reduceExistingBuys(itemKey, -delta);
            if (remaining > 0) {
                add(itemKey, null, -remaining);
            }
            return;
        }
        remaining = reduceExistingSells(itemKey, delta);
        if (remaining > 0) {
            add(itemKey, null, remaining);
        }
    }

    void resetItem(String itemKey) {
        for (int i = trades.size() - 1; i >= 0; i--) {
            if (itemKey != null && itemKey.equals(trades.get(i).getItemKey())) {
                trades.remove(i);
            }
        }
    }

    void removeExecuted(List<StockReviewPendingPurchase> executionOrder, int failedIndex) {
        if (executionOrder == null) {
            return;
        }
        for (int i = 0; i < failedIndex; i++) {
            removeMatching(executionOrder.get(i));
        }
    }

    private void removeMatching(StockReviewPendingPurchase executed) {
        if (executed == null) {
            return;
        }
        for (int i = trades.size() - 1; i >= 0; i--) {
            StockReviewPendingPurchase trade = trades.get(i);
            if (trade.matches(executed.getItemKey(), executed.getSubmarketId())
                    && trade.getQuantity() == executed.getQuantity()) {
                trades.remove(i);
                return;
            }
        }
    }

    private StockReviewPendingPurchase find(String itemKey, String submarketId) {
        for (int i = 0; i < trades.size(); i++) {
            StockReviewPendingPurchase purchase = trades.get(i);
            if (purchase.matches(itemKey, submarketId)) {
                return purchase;
            }
        }
        return null;
    }

    private int reduceExistingBuys(String itemKey, int quantity) {
        int remaining = quantity;
        for (int i = trades.size() - 1; i >= 0 && remaining > 0; i--) {
            StockReviewPendingPurchase trade = trades.get(i);
            if (!itemKey.equals(trade.getItemKey()) || !trade.isBuy()) {
                continue;
            }
            int reduced = Math.min(remaining, trade.getQuantity());
            trade.addQuantity(-reduced);
            remaining -= reduced;
            if (trade.isZero()) {
                trades.remove(i);
            }
        }
        return remaining;
    }

    private int reduceExistingSells(String itemKey, int quantity) {
        int remaining = quantity;
        for (int i = trades.size() - 1; i >= 0 && remaining > 0; i--) {
            StockReviewPendingPurchase trade = trades.get(i);
            if (!itemKey.equals(trade.getItemKey()) || !trade.isSell()) {
                continue;
            }
            int reduced = Math.min(remaining, -trade.getQuantity());
            trade.addQuantity(reduced);
            remaining -= reduced;
            if (trade.isZero()) {
                trades.remove(i);
            }
        }
        return remaining;
    }
}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/trade/StockReviewPendingTrades.kt
package weaponsprocurement.ui.stockreview.trade

import weaponsprocurement.ui.stockreview.actions.*
import weaponsprocurement.ui.stockreview.state.*
import weaponsprocurement.ui.stockreview.rows.*
import weaponsprocurement.ui.stockreview.tooltips.*
import weaponsprocurement.ui.stockreview.rendering.*
import weaponsprocurement.ui.stockreview.trade.*

import weaponsprocurement.ui.*

import java.util.ArrayList
import java.util.Collections

class StockReviewPendingTrades {
    private val trades = ArrayList<StockReviewPendingTrade>()

    fun asList(): List<StockReviewPendingTrade> = Collections.unmodifiableList(trades)

    fun isEmpty(): Boolean = trades.isEmpty()

    fun clear() {
        trades.clear()
    }

    fun replaceWith(source: List<StockReviewPendingTrade>?) {
        trades.clear()
        if (source == null) return
        for (trade in source) {
            if (!trade.isZero()) {
                val copy = trade.copy()
                if (copy != null) trades.add(copy)
            }
        }
    }

    fun add(itemKey: String?, submarketId: String?, quantity: Int) {
        if (itemKey.isNullOrEmpty() || quantity == 0) return
        val existing = find(itemKey, submarketId)
        if (existing == null) {
            val trade = StockReviewPendingTrade.create(itemKey, submarketId, quantity)
            if (trade != null) trades.add(trade)
            return
        }
        existing.addQuantity(quantity)
        if (existing.isZero()) trades.remove(existing)
    }

    fun adjustItemNet(itemKey: String?, delta: Int) {
        if (itemKey.isNullOrEmpty() || delta == 0) return
        if (delta < 0) {
            val remaining = reduceExistingBuys(itemKey, -delta)
            if (remaining > 0) add(itemKey, null, -remaining)
            return
        }
        val remaining = reduceExistingSells(itemKey, delta)
        if (remaining > 0) add(itemKey, null, remaining)
    }

    fun resetItem(itemKey: String?) {
        for (i in trades.size - 1 downTo 0) {
            if (itemKey != null && itemKey == trades[i].itemKey) {
                trades.removeAt(i)
            }
        }
    }

    fun removeExecuted(executionOrder: List<StockReviewPendingTrade>?, failedIndex: Int) {
        if (executionOrder == null) return
        for (i in 0 until failedIndex) {
            removeMatching(executionOrder[i])
        }
    }

    private fun removeMatching(executed: StockReviewPendingTrade?) {
        if (executed == null) return
        for (i in trades.size - 1 downTo 0) {
            val trade = trades[i]
            if (trade.matches(executed.itemKey, executed.submarketId) &&
                trade.quantity == executed.quantity
            ) {
                trades.removeAt(i)
                return
            }
        }
    }

    private fun find(itemKey: String, submarketId: String?): StockReviewPendingTrade? {
        for (trade in trades) {
            if (trade.matches(itemKey, submarketId)) return trade
        }
        return null
    }

    private fun reduceExistingBuys(itemKey: String, quantity: Int): Int {
        var remaining = quantity
        var i = trades.size - 1
        while (i >= 0 && remaining > 0) {
            val trade = trades[i]
            if (itemKey != trade.itemKey || !trade.isBuy()) {
                i--
                continue
            }
            val reduced = Math.min(remaining, trade.quantity)
            trade.addQuantity(-reduced)
            remaining -= reduced
            if (trade.isZero()) trades.removeAt(i)
            i--
        }
        return remaining
    }

    private fun reduceExistingSells(itemKey: String, quantity: Int): Int {
        var remaining = quantity
        var i = trades.size - 1
        while (i >= 0 && remaining > 0) {
            val trade = trades[i]
            if (itemKey != trade.itemKey || !trade.isSell()) {
                i--
                continue
            }
            val reduced = Math.min(remaining, -trade.quantity)
            trade.addQuantity(reduced)
            remaining -= reduced
            if (trade.isZero()) trades.removeAt(i)
            i--
        }
        return remaining
    }
}